llama_index/chat_engine/react.py

```diff
@@ -0,0 +1,94 @@
+from typing import Any, List, Optional
+
+from llama_index.chat_engine.types import BaseChatEngine, ChatHistoryType
+from llama_index.chat_engine.utils import to_langchain_chat_history
+from llama_index.indices.query.base import BaseQueryEngine
+from llama_index.indices.service_context import ServiceContext
+from llama_index.langchain_helpers.agents.agents import (
+    AgentExecutor,
+    AgentType,
+    initialize_agent,
+)
+from llama_index.llm_predictor.base import LLMPredictor
+from llama_index.response.schema import RESPONSE_TYPE, Response
+from llama_index.tools.query_engine import QueryEngineTool
+from langchain.memory import ConversationBufferMemory
+from langchain.memory.chat_memory import BaseChatMemory
+
+
+class ReActChatEngine(BaseChatEngine):
+    """ReAct Chat Engine.
+
+
+    Use a ReAct agent loop with query engine tools. Implemented via LangChain agent.
+    """
+
+    def __init__(
+        self,
+        query_engine_tools: List[QueryEngineTool],
+        service_context: Optional[ServiceContext] = None,
+        memory: Optional[BaseChatMemory] = None,
+        chat_history: Optional[ChatHistoryType] = None,
+        verbose: bool = False,
+    ) -> None:
+        self._query_engine_tools = query_engine_tools
+        self._service_context = service_context or ServiceContext.from_defaults()
+        if chat_history is not None and memory is not None:
+            raise ValueError("Cannot specify both memory and chat_history.")
+
+        if memory is None:
+            history = to_langchain_chat_history(chat_history)
+            memory = ConversationBufferMemory(
+                memory_key="chat_history", chat_memory=history
+            )
+
+        self._memory = memory
+        self._verbose = verbose
+
+        self._agent = self._create_agent()
+
+    @classmethod
+    def from_defaults(
+        cls,
+        query_engine: BaseQueryEngine,
```

Comment:
imo making this a list will make this more versatile to be used as standalone (instead of only from as_chat_engine)

Reply:
Fair, my initial idea is that `from_defaults` exposes a simpler, more limited constructor (i.e. one query engine, where you can pass the name, description separately). The main constructor takes list of tools, and user can use that if desired.

Reply:
Yeah but i thought from_defaults also has a dual purpose in other classes of being able to set defaults..seems like you're also doing that in the `__init__` though which seems a bit inconsistent with other classes

Reply:
fair point, will revise to be more consistent